rfd/0068-session-recording-modes.md

```diff
@@ -0,0 +1,203 @@
+---
+authors: Gabriel Corado (gabriel.oliveira@goteleport.com)
+state: draft
+---
+
+# RFD 68 - Session recording modes
+
+## What
+
+This RFD proposes giving users a configuration to decide how to deal with
+sessions that can't be recorded.
+
+## Why
+
+In cases where Teleport is configured to record sessions on the node using the
+asynchronous format, disk failures (for example, no space left) causes the node
+to be inaccessible. Where Teleport is the only way of accessing those resources,
+administrators cannot perform the necessary actions on these nodes to bring them
+back to a normal state.
+
+## Goals
+
+* Prevent nodes from being inaccessible when there are session recording
+  failures.
+* Enable users to configure the strictness of the presence of session
+  recordings.
+
+**Non-goals:**
+
+* Infrastructure-related metrics (such as disk space left).
+* Enable access when the recording mode is set to the proxy.
+* Handle audit logging (events) failures.
+* Cover applications and databases "session recordings".
+
+## Details
+
+As mentioned before, recording modes will define how Teleport proceeds in case
+of failures. We only consider asynchronous session recordings on the servers
+(`session_recording: node` on Teleport config) for this RFD. The failures
+covered are related to I/O while writing session recording events to the server
+disk.
+
+There are going to be two audit modes:
+
+* **Strict**: Teleport won't tolerate session recording to fail. In practice, if
+  Teleport cannot record the session, it will not start or will be terminated
+  (in case it is already in progress). This mode will be the recommended one as
+  it guarantees all sessions are being recorded.
+* **Best-effort**: Session recording failures are tolerated in this mode, meaning
+  that Teleport will allow sessions without recording them. System
+  administrators might use this mode to be able to access resources continually.
+
+## UX
+
+### Configuration
+
+The configuration is going to be done at the role level. The role option
+`record_session` will be extended to hold the audit mode values: `strict` and
+`best_effort`. It will have one entry per session kind (desktop, databases,
+applications, Kubernetes, and SSH). want the same behavior for all kinds.
+
+In addition, there will be a `default` entry for setting a default value used
+when sessions kind doesn't have one set.  Currently, Teleport doesn't prevent
+sessions from happening if there is a failure on the recording. To keep this
+behavior after introducing the recording modes, the default value will be set
+to `best_effort`.
```

Comment:
As discussed in the [implementation PR](https://github.com/gravitational/teleport/pull/12916#discussion_r886031101), the current behavior actually does prevent ssh sessions from functioning when session recording fails. Can we update this comment to reflect that defaulting to `best_effort` is a behavioral change?